crates/tx5/src/ep3.rs

```diff
@@ -0,0 +1,609 @@
+        match self {
+            Self::Error(err) => {
+                f.debug_struct("Error").field("err", err).finish()
+            }
+            Self::Connected { peer_url } => {
+                let url = format!("{peer_url}");
+                f.debug_struct("Connected").field("peer_url", &url).finish()
+            }
+            Self::Disconnected { peer_url } => {
+                let url = format!("{peer_url}");
+                f.debug_struct("Disconnected")
+                    .field("peer_url", &url)
+                    .finish()
+            }
+            Self::Message { peer_url, .. } => {
+                let url = format!("{peer_url}");
+                f.debug_struct("Message").field("peer_url", &url).finish()
+            }
+        }
+    }
+}
+
+/// A signal server url.
+pub type SigUrl = Tx5Url;
+
+/// A peer connection url.
+pub type PeerUrl = Tx5Url;
+
+type SigMap = HashMap<SigUrl, (u64, AbortableTimedSharedFuture<Arc<Sig>>)>;
+
+/// Callback in charge of sending preflight data if any.
+pub type PreflightSendCb = Arc<
+    dyn Fn(&PeerUrl) -> BoxFuture<'static, Result<Vec<u8>>>
+        + 'static
+        + Send
+        + Sync,
+>;
+
+/// Callback in charge of validating preflight data if any.
+pub type PreflightCheckCb = Arc<
+    dyn Fn(&PeerUrl, Vec<u8>) -> BoxFuture<'static, Result<()>>
+        + 'static
+        + Send
+        + Sync,
+>;
+
+/// Tx5 endpoint version 3 configuration.
+pub struct Config3 {
+    /// Maximum count of open connections. Default 4096.
+    pub connection_count_max: u32,
+
+    /// Max backend send buffer bytes (per connection). Default 64 KiB.
+    pub send_buffer_bytes_max: u32,
+
+    /// Max backend recv buffer bytes (per connection). Default 64 KiB.
+    pub recv_buffer_bytes_max: u32,
+
+    /// Maximum receive message reconstruction bytes in memory
+    /// (accross entire endpoint). Default 512 MiB.
+    pub incoming_message_bytes_max: u32,
+
+    /// Maximum size of an individual message. Default 16 MiB.
+    pub message_size_max: u32,
+
+    /// Default timeout for network operations. Default 60 seconds.
+    pub timeout: std::time::Duration,
+
+    /// Starting backoff duration for retries. Default 5 seconds.
+    pub backoff_start: std::time::Duration,
+
+    /// Max backoff duration for retries. Default 60 seconds.
+    pub backoff_max: std::time::Duration,
+
+    /// If the protocol should manage a preflight message,
+    /// set the callbacks here, otherwise no preflight will
+    /// be sent nor validated. Default: None.
+    pub preflight: Option<(PreflightSendCb, PreflightCheckCb)>,
+}
+
+impl Default for Config3 {
+    fn default() -> Self {
+        Self {
+            connection_count_max: 4096,
+            send_buffer_bytes_max: 64 * 1024,
+            recv_buffer_bytes_max: 64 * 1024,
+            incoming_message_bytes_max: 512 * 1024 * 1024,
+            message_size_max: 16 * 1024 * 1024,
```

Comment:
What's the difference between the message side and the incoming message size? Just looking at this it seems unnecessary to allow more incoming than should be sent?

Reply:
```rust
    /// Maximum receive message reconstruction bytes in memory
    /// (accross entire endpoint). Default 512 MiB.
    pub incoming_message_bytes_max: u32,

    /// Maximum size of an individual message. Default 16 MiB.
    pub message_size_max: u32,
```

The `incoming_message_bytes_max` is our total buffer for reconstructing multi-part/chunked messages. It is shared across all connections, and allows us to keep many connections open without worrying about their memory usage.

The `message_size_max` is literally the biggest message we allow to be sent to us.